### PR TITLE
Add an analyzer for analyzing required vs optional dependencies.

### PR DESCRIPTION
**Pull request by ljcollins25**

### CODE REVIEW

File: Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/Simulator/BuildSimulator.cs [hunk 1]
```diff
@@ -62,23 +62,26 @@ internal sealed class BuildSimulatorAnalyzer : Analyzer
...
 
+        public string ResultsDirectory;
+
+
         public bool[] IncludedExecutionSteps = new bool[EnumTraits<PipExecutionStep>.MaxValue + 1];
-        private static string resultsFormat;
 
-        private readonly PipExecutionData m_executionData = new PipExecutionData();
+        public readonly PipExecutionData ExecutionData;
+        public int Increment { get; set; } = 12;
```
Comment (Kr1sso) [inline Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/Simulator/BuildSimulator.cs]: Nit: You noted you want to have this settable via a command line option?

File: Public/Src/Utilities/Utilities/Tracing/BinaryLogReader.cs [hunk 2]
```diff
@@ -248,7 +248,9 @@ protected uint ReadPathEvent(EventReader reader)
             AbsolutePath parent = reader.ReadAbsolutePath();
             int index = reader.ReadInt32Compact();
             string name = reader.ReadString();
-            AbsolutePath path = parent.IsValid ? parent.Combine(m_context.PathTable, PathAtom.Create(m_context.StringTable, name)) : AbsolutePath.Create(m_context.PathTable, name);
+            AbsolutePath path = parent.IsValid 
+                ? parent.Combine(m_context.PathTable, PathAtom.UnsafeCreateFrom(StringId.Create(m_context.StringTable, name))) 
+                : AbsolutePath.Create(m_context.PathTable, name);
```
Comment (Kr1sso) [inline Public/Src/Utilities/Utilities/Tracing/BinaryLogReader.cs]: Is this the macOS specific fix when running the analysis on Windows? If so, could you leave a comment there so we can look at this later potentially when addressing other x-plat issues too?

File: Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/Simulator/BuildSimulator.cs [hunk 3]
```diff
@@ -62,23 +62,26 @@ internal sealed class BuildSimulatorAnalyzer : Analyzer
         /// </summary>
         public string OutputDirectory;
 
+        public string ResultsDirectory;
```
Comment (kewilli) [inline Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/Simulator/BuildSimulator.cs]: Nit: Remove empty line

File: Public/Src/Tools/Execution.Analyzer/DisplayTable.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                for (int i = 0; i < row.Length; i++)
+                {
+                    var value = row[i] ?? string.Empty;
+                    sb.Append(' ', m_maxColumnLengths[i] - value.Length);
+                    sb.Append(value);
+                    if (i != (row.Length - 1))
+                    {
+                        sb.Append(m_columnDelimeter);
+                    }
+                }
```
Comment (dannyvv) [inline Public/Src/Tools/Execution.Analyzer/DisplayTable.cs]: String already has handy formatter structures for this....
[https://stackoverflow.com/questions/2978311/format-a-string-into-columns](https://stackoverflow.com/questions/2978311/format-a-string-into-columns)

Comment (ljcollins25) [reply Public/Src/Tools/Execution.Analyzer/DisplayTable.cs]: That just adds fixed-width padding. It ends up being more work if you dynamically decide the padding level at runtime.

---
In reply to: [293125956](https://github.com/microsoft/BuildXL/pull/444#discussion_r293125956) [](ancestors = 293125956)

**Pull request merged**